### PR TITLE
Support for Touch Alarm

### PR DESCRIPTION
**Pull request by microDev1**
This PR adds support for setting a Touch Alarm.

**Port(s) supported:** _(as of this PR)_
- esp32s2

**Test Code:**
```python
import time
import alarm
import digitalio
from board import *

led = digitalio.DigitalInOut(LED)
led.direction = digitalio.Direction.OUTPUT
led.value = True

time.sleep(5)

print("WakeAlarm :- ", alarm.wake_alarm)
print()

alarm.exit_and_deep_sleep_until_alarms(alarm.touch.TouchAlarm(IO1))
```

### CODE REVIEW

File: ports/esp32s2/common-hal/alarm/__init__.c [hunk 1]
```diff
@@ -110,6 +115,17 @@ STATIC void _setup_sleep_alarms(bool deep_sleep, size_t n_alarms, const mp_obj_t
...
+            if (!touch_alarm_set) {
+                if (deep_sleep) {
+                    alarm_touch_touchalarm_set_alarm(MP_OBJ_TO_PTR(alarms[i]));
+                    touch_alarm_set = true;
+                } else {
+                    mp_raise_NotImplementedError(translate("TouchAlarm not available in light sleep"));
+                }
+            } else {
+                mp_raise_ValueError(translate("Only one alarm.touch alarm can be set."));
+            }
```
Comment (tannewt) [inline ports/esp32s2/common-hal/alarm/__init__.c:128]: Please factor this out into the touch alarm code. (See my pin alarm PR for how I've done it there.)

File: ports/esp32s2/common-hal/alarm/touch/TouchAlarm.c [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    // wait for touch data to reset
+    mp_hal_delay_ms(10);
+
+    uint32_t touch_value;
+    touch_pad_sleep_channel_read_smooth(touch_channel, &touch_value);
+    touch_pad_sleep_set_threshold(touch_channel, touch_value * 0.1); //10%
+}
+
+bool alarm_touch_touchalarm_woke_us_up(void) {
+    return false;
```
Comment (tannewt) [inline ports/esp32s2/common-hal/alarm/touch/TouchAlarm.c:93]: Does this work when pretending to deep sleep? Generally, you can use an internal callback to notify the main CP task to wake back up.

**Review by tannewt — DISMISSED** [review]
Thank you for the PR! Just a couple comments.

I wish my pin alarm PR was already in. I heard you hit the same issue I did. I hope we can figure it out soon.

File: ports/esp32s2/common-hal/alarm/touch/TouchAlarm.c (hunk 2 above)
Comment (microDev1) [reply ports/esp32s2/common-hal/alarm/touch/TouchAlarm.c:93]: Now using touch interrupt for `pretend-to-sleep`.

Comment (tannewt) [reply ports/esp32s2/common-hal/alarm/touch/TouchAlarm.c:93]: I don't see any new changes. Did you forget to push them?

File: ports/esp32s2/common-hal/alarm/touch/TouchAlarm.c [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    alarm_touch_touchalarm_obj_t *touch_alarm = MP_OBJ_NULL;
+
+    for (size_t i = 0; i < n_alarms; i++) {
+        if (MP_OBJ_IS_TYPE(alarms[i], &alarm_touch_touchalarm_type)) {
+            if (!touch_alarm_set) {
+                if (deep_sleep) {
+                    touch_alarm = MP_OBJ_TO_PTR(alarms[i]);
+                    touch_alarm_set = true;
+                } else {
+                    mp_raise_NotImplementedError(translate("TouchAlarm not available in light sleep"));
```
Comment (tannewt) [inline ports/esp32s2/common-hal/alarm/touch/TouchAlarm.c:91]: Does the interrupt work during light sleep?

Comment (microDev1) [reply ports/esp32s2/common-hal/alarm/touch/TouchAlarm.c]: Thanks for pointing this out. IDF implementation of `TouchWake` doesn't work for `LightSleep`. 
I forgot to test this with mechanics of `pretend-to-sleep`... wake on interrupt from light sleep is working now.

**Review by jepler — APPROVED** [review]
@tannewt 's requested changes seem to have been addressed.  Merging so we can have this in beta3.

**Comment by dglaude** [comment]
I tested this feature with the ESP32S2.
It seems that my board wakeup twice in a raw when I touch.
Basically it made this code blink the blue LED twice in a raw for 5 seconds (even with a single touch of the pin).
This is tested with 6.1.0-beta3.

**Comment by dglaude** [comment]
Ok, the double trigger on touch only seems to occur with the USB connected to computer, but not when on a power supply. So that would be on fake sleep path.

Also wake-up speed on touch seems much faster on fake sleep than on real sleep... understandable but should it be better faked?

**Comment by microDev1** [comment]
Hi @dglaude Thanks for testing this out. Can you please open an issue for this. :-)

**Pull request merged**